pkg/kvm/kvm_test.go

```diff
@@ -0,0 +1,438 @@
+		}
+
+		s.write(s.greeting, conn)
+		s.handleCall(s.capabilitiesNegotiation, conn)
+		for _, call := range s.calls {
+			s.handleCall(call, conn)
+			atomic.AddUint32(&s.callIndex, 1)
+		}
+	}()
+
+	return s
+}
+
+func (s *mockQmpServer) Stop() {
+	if s.socket != nil {
+		if err := s.socket.Close(); err != nil {
+			log.Panicf("Failed to close socket: %v", err)
+		}
+	}
+	if err := os.RemoveAll(s.testDir); err != nil {
+		log.Panicf("Failed to delete test dir: %v", err)
+	}
+}
+
+func (s *mockQmpServer) ExpectAddChardev(id string) *mockQmpServer {
+	s.calls = append(s.calls, mockCall{
+		response: `{"return": {"pty": "/tmp/dev/pty/42"}}` + "\n",
+		expectedArgs: []string{
+			`"execute":"chardev-add"`,
+			`"id":"` + id + `"`,
+			`"path":"` + filepath.Join(s.testDir, id) + `"`,
+		},
+	})
+	return s
+}
+
+func (s *mockQmpServer) ExpectAddVirtioBlk(id string, chardevID string) *mockQmpServer {
+	s.calls = append(s.calls, mockCall{
+		response: genericQmpOk,
+		expectedArgs: []string{
+			`"execute":"device_add"`,
+			`"driver":"vhost-user-blk-pci"`,
+			`"id":"` + id + `"`,
+			`"chardev":"` + chardevID + `"`,
+		},
+	})
+	return s
+}
+
+func (s *mockQmpServer) ExpectDeleteChardev(id string) *mockQmpServer {
+	s.calls = append(s.calls, mockCall{
+		response: genericQmpOk,
+		expectedArgs: []string{
+			`"execute":"chardev-remove"`,
+			`"id":"` + id + `"`,
+		},
+	})
+	return s
+}
+
+func (s *mockQmpServer) ExpectDeleteVirtioBlkWithEvent(id string) *mockQmpServer {
+	s.ExpectDeleteVirtioBlk(id)
+	s.calls[len(s.calls)-1].event =
+		`{"event":"DEVICE_DELETED","data":{"path":"/some/path","device":"` +
+			id + `"},"timestamp":{"seconds":1,"microseconds":2}}` + "\n"
+	return s
+}
+
+func (s *mockQmpServer) ExpectDeleteVirtioBlk(id string) *mockQmpServer {
+	s.calls = append(s.calls, mockCall{
+		response: genericQmpOk,
+		expectedArgs: []string{
+			`"execute":"device_del"`,
+			`"id":"` + id + `"`,
+		},
+	})
+	return s
+}
+
+func (s *mockQmpServer) WithErrorResponse() *mockQmpServer {
+	if len(s.calls) == 0 {
+		log.Panicf("No instance to add a QMP error")
+	}
+	s.calls[len(s.calls)-1].response = genericQmpError
+	return s
+}
+
+func (s *mockQmpServer) WereExpectedCallsPerformed() bool {
+	return int(atomic.LoadUint32(&s.callIndex)) == len(s.calls)
+}
+
+func (s *mockQmpServer) handleCall(call mockCall, conn net.Conn) {
+	req := s.read(conn)
+	for _, expectedArg := range call.expectedArgs {
+		if !strings.Contains(req, expectedArg) {
+			s.test.Errorf("Expected to find argument %v in %v", expectedArg, req)
+		}
+	}
+	s.write(call.response, conn)
+	if call.event != "" {
+		time.Sleep(time.Millisecond * 1)
+		s.write(call.event, conn)
+	}
+}
+
+func (s *mockQmpServer) write(data string, conn net.Conn) {
+	log.Println("QMP server send:", data)
+	_, err := conn.Write([]byte(data))
+	if err != nil {
+		log.Panicf("QMP server failed to write: %v", data)
+	}
+}
+
+func (s *mockQmpServer) read(conn net.Conn) string {
+	buf := make([]byte, 512)
```

Comment:
looks like this is new style of testing, please consider either change the rest of tests to look the same or keep the old style... it is hard to maintain project that has a lot of coding styles in it..

Reply:
from slack: I'm referring to `Expect*` calls
I'm not against it, I just looking for the same look and feel across the project

Reply:
sure